julia/src/native_function.jl

```diff
@@ -0,0 +1,277 @@
+
+        assigment = :($(esc(name)) = $(native_func))
+        if 'c' in flags
+            assigment = Expr(:const, assigment)
+        end
+        return assigment
+    end
+end
+
+
+"holds information on how to map an argument argment in a signature into a julia type"
+struct Binder
+    name::Symbol
+    type  # Intentionally left unrestricted for now. No restriction seems useful, it is either one of half a dozen scalar types or an ArrayBuilder
+    implicit::Bool
+end
+
+"Creates an `Array{T,N}``, if all elements of size are present then that array is initialized to hold `size`"
+struct ArrayBuilder{T,N}
+    size::Vector{Union{Symbol,Int}}
+end
+ArrayBuilder{T}(size) where T = ArrayBuilder{T,length(size)}(size)
+
+
+
+"""
+    NativeFunction{R}
+
+Callable Julia wrapper of some native Dex function.
+Returns type from calling it is is `R`.
+
+Usually constructed using [`@dex_func_str`](@ref),
+or via `NativeFunction(atom)` on some [`DexCall.Atom`](@ref).
+"""
+struct NativeFunction{R} <: Function
+    c_func_ptr::Ptr{Nothing}
+    implict_argument_signature::Vector{Binder}
+    explicit_argument_signature::Vector{Binder}
+    result_signature::Vector{Binder}
+end
+
+NativeFunction(atom::Atom, jit=JIT) = NativeFunction(atom.ptr, atom.ctx, jit)
+function NativeFunction(atom::Ptr{HsAtom}, ctx=PRELUDE, jit=JIT)
+    c_func_ptr = compile(ctx, atom, jit)
+    sig_ptr = get_function_signature(c_func_ptr, jit)
+    sig_ptr == C_NULL && error("Failed to retrieve the function signature")
+
+    try
+        signature = NativeFunctionSignatureJL(sig_ptr)
+        argument_signature = parse_sig(signature.arg)
+        result_signature = parse_sig(signature.res)
+        R = result_type(result_signature)
+        f = NativeFunction{R}(
+            c_func_ptr,
+            filter(x->x.implicit, argument_signature),
+            filter(x->!x.implicit, argument_signature),
+            result_signature
+        )
+
+        # We are attaching this to f.result_signature because only mutable types can have finalizers
+        finalizer(f.result_signature) do _
+            unload(c_func_ptr, jit)
+        end
+        return f
+    catch
+        unload(c_func_ptr, jit)
+        rethrow()
+    finally
+        free_function_signature(sig_ptr)
+    end
+end
+
+function (f::NativeFunction{R})(args...)::R where R
+    if length(args) != length(f.explicit_argument_signature)
+        throw(ArgumentError("wrong number of inputs, expected: $(length(f.explicit_argument_signature) )"))
+    end
+
+    named_sizes = Dict{Symbol, Int}()
+    arg_ctypes = map(f.explicit_argument_signature, args) do binder, arg
+        to_ctype!(named_sizes, binder.type, arg)
+    end
+
+    implict_values = map(f.implict_argument_signature) do binder
+        # Right now the Exported API only supports implicts that are sizes
+        # and now all of them will have been determined by the preverious `to_ctype!` step
+        @assert binder.type == Int32
```

Comment:
This is not important.
It is just an assertion about my belief about the world.
But AFAICT is is always true for things that an be called in the API?
And the other code is invalid if it isn't true?
If it isn't true I would like a example that  shows it is not to add to the tests.

Reply:
At the moment it might be true, but I expect that there will come a time where it won't (e.g. Int64 might be another valid option). I'd rather not hard-code any assumptions that are unnecessary and we don't make in Dex itself.

Reply:
```suggestion
```